src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java

```diff
@@ -1230,20 +1192,31 @@ public Queue.Item itemInQueue() {
 
     @Extension
     public static class RunningTasks {
-        /** keys are {@link PlaceholderTask#cookie}s */
-        private final Map<String, RunningTask> runningTasks = new HashMap<>();
+        private final Map<StepContext, RunningTask> runningTasks = new HashMap<>();
 
-        synchronized <T> T withRunningTasks(Function<Map<String, RunningTask>, T> fn) {
-            return fn.apply(runningTasks);
+        private static RunningTask find(StepContext context) {
+            RunningTasks holder = ExtensionList.lookupSingleton(RunningTasks.class);
+            synchronized (holder) {
+                return holder.runningTasks.computeIfAbsent(context, k -> new RunningTask());
+            }
         }
 
-        synchronized void withRunningTasks(Consumer<Map<String, RunningTask>> fn) {
-            fn.accept(runningTasks);
+        static <T> T get(StepContext context, Function<RunningTask, T> fn) {
+            return fn.apply(find(context));
         }
 
-        static RunningTasks get() {
-            return ExtensionList.lookupSingleton(RunningTasks.class);
+        static void run(StepContext context, Consumer<RunningTask> fn) {
+            fn.accept(find(context));
         }
+
+        static RunningTask remove(StepContext context) {
+            RunningTasks holder = ExtensionList.lookupSingleton(RunningTasks.class);
+            synchronized (holder) {
+                RunningTask t = holder.runningTasks.remove(context);
+                return t != null ? t : new RunningTask();
```

Comment:
Do we hit the `t == null` condition here in normal cases? I wonder whether it would be worth logging a FINE message in that case for future debugging.